src/NHibernate.Test/NHSpecificTest/NH2420/Fixture.cs

```diff
@@ -55,41 +56,56 @@ public void ShouldBeAbleToReleaseSuppliedConnectionAfterDistributedTransaction()
 		{
 			string connectionString = FetchConnectionStringFromConfiguration();
 			ISession s;
-			using (var ts = new TransactionScope())
+			DbConnection connection = null;
+			try
 			{
-				// Enlisting DummyEnlistment as a durable resource manager will start
-				// a DTC transaction
-				System.Transactions.Transaction.Current.EnlistDurable(
-					DummyEnlistment.Id,
-					new DummyEnlistment(),
-					EnlistmentOptions.None);
+				using (var ts = new TransactionScope())
+				{
+					// Enlisting DummyEnlistment as a durable resource manager will start
+					// a DTC transaction
+					System.Transactions.Transaction.Current.EnlistDurable(
+						DummyEnlistment.Id,
+						new DummyEnlistment(),
+						EnlistmentOptions.None);
 
-				DbConnection connection;
-				if (sessions.ConnectionProvider.Driver.GetType() == typeof(OdbcDriver))
-					connection = new OdbcConnection(connectionString);
-				else
-					connection = new SqlConnection(connectionString);
+					if (sessions.ConnectionProvider.Driver.GetType() == typeof(OdbcDriver))
+						connection = new OdbcConnection(connectionString);
+					else
+						connection = new SqlConnection(connectionString);
 
-				using (connection)
-				{
 					connection.Open();
 					using (s = Sfi.OpenSession(connection))
 					{
-						s.Save(new MyTable { String = "hello!" });
+						s.Save(new MyTable {String = "hello!"});
 					}
-					connection.Close();
-				}
+					// The ts disposal may try to flush the session, which, depending on the native generator
+					// implementation for current dialect, may have something to do and will then try to use
+					// the supplied connection. dispose connection here => flaky test, failing for dialects
+					// not mandating an immediate insert on native generator save.
+					// Delaying the connection disposal to after ts disposal.
 
-				ts.Complete();
+					ts.Complete();
+				}
+			}
+			finally
+			{
+				connection?.Dispose();
 			}
 
+			// It appears neither the second phase of the 2PC nor TransactionCompleted 
+			// event are guaranteed to be executed before exiting transaction scope disposal.
+			// When having only 2PC, the second phase tends to occur after reaching that point
+			// here. When having TransactionCompleted event, this event and the second phase
+			// tend to occur before reaching here. But some other NH cases demonstrate that
+			// TransactionCompleted may also occur "too late".
+			((ISessionImplementor) s).TransactionContext?.WaitOne();
```

Comment:
Probably not the first time... Maybe should we remove all the other alike catches which still liter the tests.